File: workshop/src/lib/validation/type-validator.ts
import type { ResourceType, Diagnostic, SchemaRegistrySnapshot } from '../types';

export function validateType(
  type: ResourceType,
  snapshot: SchemaRegistrySnapshot,
): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];

  // Check duplicate field names
  const fieldNames = type.fields.map(f => f.name);
  const seen = new Set<string>();
  for (const name of fieldNames) {
    if (seen.has(name)) {
      diagnostics.push({ severity: 'error', message: `Duplicate field name: "${name}"` });
    }
    seen.add(name);
  }

  // Check relationship target types exist
  for (const rel of type.relationships) {
    if (!snapshot.types.has(rel.targetType)) {
      diagnostics.push({
        severity: 'error',
        message: `Relationship "${rel.label}" references non-existent type "${rel.targetType}"`,
      });
    }
  }

  // Warning: orphan nodes (states with no incoming edges, excluding the initial state)
  const allStates = new Set(type.stateMachine.flatMap(t => [t.from, t.to]));
  const initialStates = new Set(type.stateMachine.map(t => t.from).filter(
    from => !type.stateMachine.some(t => t.to === from)
  ));
  const hasIncoming = new Set(type.stateMachine.map(t => t.to));
  for (const state of allStates) {
    if (!hasIncoming.has(state) && !initialStates.has(state)) {
      diagnostics.push({
        severity: 'warning',
        message: `State "${state}" has no incoming transitions (orphan node)`,
      });
    }
  }

  // Warning: no abilities declared
  if (type.abilities.length === 0) {
    diagnostics.push({
      severity: 'warning',
      message: `Type "${type.name}" has no abilities declared — it cannot be queried or modified`,
    });
  }

  // Info: PascalCase naming convention
  if (type.name[0] !== type.name[0].toUpperCase()) {
    diagnostics.push({
      severity: 'info',
      message: `Type name "${type.name}" should use PascalCase by convention`,
    });
  }

  return diagnostics;
}

File: heirloom-server/src/main/java/com/heirloom/schema/service/TypeValidator.java
package com.heirloom.schema.service;

import com.heirloom.schema.domain.*;
import com.heirloom.schema.repository.ResourceTypeRepository;

import java.util.*;

/**
 * Validates a Resource Type definition against Heirloom's semantic rules.
 * Mirrors the TypeScript type-validator.ts from the Workshop frontend.
 *
 * <p>Key rules:
 * <ul>
 *   <li>Field name uniqueness within a type</li>
 *   <li>Relationship targets must exist in the registry</li>
 *   <li>Orphan state nodes (no incoming transitions) produce a warning</li>
 *   <li>Empty abilities list produces a warning</li>
 *   <li>PascalCase naming convention (info-level)</li>
 *   <li>Initial state required when state machine is non-empty (I-0)</li>
 *   <li>Initial state must be a valid state in the state machine (I-0)</li>
 * </ul>
 */
public class TypeValidator {

    /**
     * Validates a resource type definition. Returns diagnostics
     * keyed by severity level.
     */
    public static List<Diagnostic> validate(ResourceType type,
                                            Map<String, ResourceType> knownTypes) {
        List<Diagnostic> diagnostics = new ArrayList<>();

        checkDuplicateFields(type, diagnostics);
        checkRelationshipTargets(type, knownTypes, diagnostics);
        checkOrphanStates(type, diagnostics);
        checkAbilities(type, diagnostics);
        checkNaming(type, diagnostics);
        checkInitialState(type, diagnostics);

        return diagnostics;
    }

    private static void checkDuplicateFields(ResourceType type, List<Diagnostic> out) {
        Set<String> seen = new HashSet<>();
        for (Field field : type.getFields()) {
            if (!seen.add(field.name())) {
                out.add(new Diagnostic(Severity.ERROR,
                        "Duplicate field name: \"" + field.name() + "\""));
            }
        }
    }

    private static void checkRelationshipTargets(ResourceType type,
                                                  Map<String, ResourceType> knownTypes,
                                                  List<Diagnostic> out) {
        for (Relationship rel : type.getRelationships()) {
            if (!knownTypes.containsKey(rel.targetType())) {
                out.add(new Diagnostic(Severity.ERROR,
                        "Relationship \"" + rel.label()
                        + "\" references non-existent type \""
                        + rel.targetType() + "\""));
            }
        }
    }

    private static void checkOrphanStates(ResourceType type, List<Diagnostic> out) {
        Set<String> allStates = new HashSet<>();
        Set<String> hasIncoming = new HashSet<>();
        Set<String> initialStates = new HashSet<>();

        for (StateTransition t : type.getStateMachine()) {
            allStates.add(t.from());
            allStates.add(t.to());
            hasIncoming.add(t.to());
        }

        // States that appear as 'from' but never as 'to' are initial states
        for (StateTransition t : type.getStateMachine()) {
            if (!hasIncoming.contains(t.from())) {
                initialStates.add(t.from());
            }
        }

        for (String state : allStates) {
            if (!hasIncoming.contains(state) && !initialStates.contains(state)) {
                out.add(new Diagnostic(Severity.WARNING,
                        "State \"" + state
                        + "\" has no incoming transitions (orphan node)"));
            }
        }
    }

    private static void checkAbilities(ResourceType type, List<Diagnostic> out) {
        if (type.getAbilities().isEmpty()) {
            out.add(new Diagnostic(Severity.WARNING,
                    "Type \"" + type.getName()
                    + "\" has no abilities declared — "
                    + "it cannot be queried or modified"));
        }
    }

    private static void checkNaming(ResourceType type, List<Diagnostic> out) {
        String name = type.getName();
        if (name != null && !name.isEmpty() && !Character.isUpperCase(name.charAt(0))) {
            out.add(new Diagnostic(Severity.INFO,
                    "Type name \"" + name
                    + "\" should use PascalCase by convention"));
        }
    }

    private static void checkInitialState(ResourceType type, List<Diagnostic> out) {
        boolean hasStateMachine = type.getStateMachine() != null && !type.getStateMachine().isEmpty();
        String initialState = type.getInitialState();

        if (hasStateMachine && (initialState == null || initialState.isBlank())) {
            out.add(new Diagnostic(Severity.ERROR,
                    "Type \"" + type.getName()
                    + "\" has a state machine but no initialState declared. "
                    + "Declare an initialState that is a valid state in the state machine."));
            return;
        }

        if (initialState != null && !initialState.isBlank() && hasStateMachine) {
            Set<String> allStates = new HashSet<>();
            for (StateTransition t : type.getStateMachine()) {
                allStates.add(t.from());
                allStates.add(t.to());
            }
            if (!allStates.contains(initialState)) {
                out.add(new Diagnostic(Severity.ERROR,
                        "Initial state \"" + initialState
                        + "\" is not a valid state in type \"" + type.getName()
                        + "\" state machine. Valid states: " + allStates));
            }
        }
    }

    // --- Diagnostic nested types ---

    public enum Severity { ERROR, WARNING, INFO }

    public record Diagnostic(Severity severity, String message) {
        public String getField() { return null; }
    }
}
